File: 23-merge-k-sorted-lists/merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
struct compare{
    bool operator()(ListNode* a, ListNode* b){
        return a->val>b->val;
    }
};
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        int k= (int)lists.size();
        if(k==0) return NULL;
        priority_queue<ListNode*, vector<ListNode*>, compare> pq;
        for(int i=0;i<k;++i){
            if(lists[i]) pq.push(lists[i]);
        }
        if(pq.empty()) return NULL;
        ListNode *temp=NULL, *last=NULL, *curr=NULL;
        while(!pq.empty()){
            curr=pq.top();
            pq.pop();
            if(curr->next) pq.push(curr->next);
            if(!last){
                last= temp= curr;
            }
            else{
                temp->next= curr;
                temp=temp->next;
            }
        }
        if(temp) temp->next=NULL;
        return last;
    }
};

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode mergeList(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode();
        ListNode tail = dummy;

        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                tail.next = l1;
                l1 = l1.next;
            } else {
                tail.next = l2;
                l2 = l2.next;
            }
            tail = tail.next;
        }
        if (l1 != null) {
            tail.next = l1;
        }
        if (l2 != null) {
            tail.next = l2;
        }
        return dummy.next;
    }
    public ListNode mergeKLists(ListNode[] lists) {
        int n= lists.length;
        if(n==0) return null;
        for(int i=n;i>1;i=(i+1)/2){
            int lo=0, hi=i-1;
            while(lo<hi){
                lists[lo]= mergeList(lists[lo],lists[hi]);
                lo++; hi--;
            }
        }
        return lists[0];
    }
}